### ISSUE TITLE
Accept linestring geometries containing a ClosePath command

### ISSUE THREAD

**Issue by anandthakker** [mapbox/vtzero#39]
Refs https://github.com/mapbox/mapbox-gl-native/issues/11579
Fixture in mvt-fixtures: https://github.com/mapbox/mvt-fixtures/pull/94

It looks like `mapbox/vector-tile` effectively just [reads `ClosePath` as `MoveTo(first_point_in_this_line)`](https://github.com/mapbox/vector-tile/blob/8aa13db7bf5087a6e557734f5fe4ceb6e95dd572/include/mapbox/vector_tile.hpp#L312-L316).  I suppose if we want mapbox-gl-native to behave _exactly_ the same way after moving from `vector-tile` to `vtzero`, we'd want to do the same thing here, but I'm not sure if that's necessary.  If, for example, simply ignoring the `ClosePath` command made more sense from vtzero's point of view, mapbox-gl-native would be able to be backwards compatible with a slight degradation that, IMO, would be acceptable.

**Comment by joto** [comment]
The important question is: Was this tile added as an example of a "usual tile" or was it added as an example for a "slightly malformed tile" that should still be accepted. Only the first case it is likely that there are actual tiles like that in the wild.

**Comment by anandthakker** [comment]
Right.  Looking at its history, I have seen no reason to think that it was intended to cover an edge case (if it were, I would have expected to see _something_ to suggest that: a code or commit comment, it originally being added in an obvious edge-case unit test, etc.).  That, and the fact that it looks "mapboxy", makes me think that it very well could have been a typical tile at the time.

**Comment by flippmoke** [comment]
I do not think we should be handling the case in which we have a `ClosePath` in a linestring or point with out error. The v2 version of the specification is very clear what is [allowed in a linestring](https://github.com/mapbox/vector-tile-spec/tree/master/2.1#4343-linestring-geometry-type). 

In [mapnik-vector-tile currently](https://github.com/mapbox/mapnik-vector-tile/blob/master/src/vector_tile_geometry_decoder.ipp#L215-L227), if it is a v1 tile we simply ignore the `closepath` as well -- but instead we just `break` decoding the entire geom at that point. However, I think at some point we need to stop designing around such edge cases. Despite V1 being very vague, I do not think the intent was ever to support `ClosePath` in linestrings and I am not sure if any encoder ever had this as a design decision (other then an encoding bug).

**Comment by anandthakker** [comment]
Options:
1. DON'T support this anomaly in vtzero, and DON'T port mapbox-gl-native to vtzero.  Ensures backwards compatibility, but deprives vtzero of a client and leaves gl-native with tech debt.
2. DON'T support this anomaly in vtzero, but DO port mabpox-gl-native to vtzero.  Avoids tech debt, but potentially breaks backwards compatilibity: any users with an old offline database containing features with this anomaly will no longer see the offending feature in their maps.
3. DO support this anomaly in vtzero and DO port mapbox-gl-native to vtzero. Ensures backwards compatibility, but leaves vtzero with tech debt.

**Comment by flippmoke** [comment]
@anandthakker Do we actually know of any tiles that do this? We ported the vast majority of our own tilesets to v2 a very long time ago, even before that we were not creating V1 tiles with closepath. Old customer data **might** contain such a situation, but I very much doubt that it is many of them. Additionally, this would only be a problem for individuals who updated their GL client and were still using very old data. 

Overall, I feel that we are pretty safe here. However, I think this continues to pop up that we do not have any tooling to check situations such this across all of our tiles that have been uploaded. I know that long ago @lbud had some long running scripts to test for issues such as this when we were updating to v2. I have no idea if they would still work or could be modified to test for something such as this.

**Comment by anandthakker** [comment]
The only tile I _know_ of that does this is the test fixture I referred to above.  I just have no idea how to estimate the likelihood that such a tile is also sitting in either (a) a non-Mapbox server or (b) on an end-user's phone in their offline cache.  cc @jfirebaugh @kkaefer

**Comment by lbud** [comment]
> I know that long ago @lbud had some long running scripts to test for issues such as this when we were updating to v2. I have no idea if they would still work or could be modified to test for something such as this.

This was a [checker](https://github.com/mapbox/checker) stack ([script](https://github.com/mapbox/checker/blob/master/lib/simplevalid.js)). I…don't remember how that worked really and would doubt it still runs as-is on our platform infrastructure 😕

**Comment by anandthakker** [comment]
Alright, I finally managed to track down the history of the offending tile:
 - Confirmed that it's a Mapbox Streets v6 vector tile, went into production on April 2, 2015.
 - Also confirmed that the closepath issue is _not_ present in the subsequent version of Mapbox Streets v6 that was deployed on May 14, 2015.
 - Furthermore, https://github.com/mapbox/mapnik-vector-tile/pull/94, an extensive PR improving mapnik-vector-tile's geometry encoding, was merged on April 29, 2015, just before the above deploy.  

Therefore, it seems fairly likely that this issue was due to an encoding issue that was fixed in mapnik . in mid-2015.  Since offline didn't [ship in native until 2016](https://github.com/mapbox/mapbox-gl-native/pull/3715), I think it's reasonably safe to assume that we can punt on supporting this issue.
